pkg/controller/worker/machineset_vmo.go

```diff
@@ -172,6 +169,75 @@ func cleanupOrphanVMODependencies(ctx context.Context, client azureclient.Vmss,
 	return nil
 }
 
+func (w *workerDelegate) determineWorkerPoolVmoDependency(ctx context.Context, infrastructureStatus *azureapi.InfrastructureStatus, workerStatus *azureapi.WorkerStatus, workerPoolName string) (*azureapi.VmoDependency, error) {
+	if !azureapihelper.IsVmoRequired(infrastructureStatus) {
+		return nil, nil
+	}
+
+	// First: Lookup the vmo dependency for the worker pool in the worker status.
+	var dependencyInStatus *azureapi.VmoDependency
+	for _, dep := range workerStatus.VmoDependencies {
+		if dep.PoolName != workerPoolName {
+			continue
+		}
+		if dependencyInStatus != nil {
+			return nil, fmt.Errorf("found more then one vmo dependencies for workerpool %s in the worker provider status", workerPoolName)
+		}
+		runVar := dep
+		dependencyInStatus = &runVar
+	}
+	if dependencyInStatus != nil {
+		return dependencyInStatus, nil
+	}
+
+	// Second: The vmo dependency was not found in the worker status. Check if a corresponding vmo exists on Azure.
+	vmoClient, err := w.clientFactory.Vmss(ctx, w.worker.Spec.SecretRef)
+	if err != nil {
+		return nil, err
+	}
+
+	vmoListAll, err := vmoClient.List(ctx, infrastructureStatus.ResourceGroup.Name)
+	if err != nil {
+		return nil, err
+	}
+	vmoList := filterGardenerManagedVmos(vmoListAll)
+
+	var existingVmo *compute.VirtualMachineScaleSet
+	for _, vmo := range vmoList {
+		if vmo.Name != nil && strings.Contains(*vmo.Name, workerPoolName) {
+			if existingVmo != nil {
+				return nil, fmt.Errorf("found multiple vmos for workerpool %q in resource group %q", workerPoolName, infrastructureStatus.ResourceGroup.Name)
+			}
+			runVar := &vmo
+			existingVmo = runVar
+		}
+	}
+	if existingVmo != nil {
+		existingVmoDependency := generateVmoDependency(existingVmo, workerPoolName)
+		workerStatus.VmoDependencies = append(workerStatus.VmoDependencies, *existingVmoDependency)
+		if err := w.updateWorkerProviderStatus(ctx, workerStatus); err != nil {
+			return nil, err
+		}
+		return existingVmoDependency, nil
+	}
+
+	// Third: No vmo for the worker pool was found on Azure. Need to create it.
+	faultDomainCount, err := azureapihelper.FindDomainCountByRegion(w.cloudProfileConfig.CountFaultDomains, w.worker.Spec.Region)
+	if err != nil {
+		return nil, err
+	}
+
+	newDependency, err := generateAndCreateVmo(ctx, vmoClient, workerPoolName, infrastructureStatus.ResourceGroup.Name, w.worker.Spec.Region, faultDomainCount)
+	if err != nil {
+		return nil, err
+	}
```

Comment:
Do we actually need to attempt to create the VMO here ? Wouldn't be enough if we just returned an error and waited for the next worker reconciliation instead of duplicating the creation logic ?

Reply:
If the Shoot is already in deletion phase then no next reconciliation flow will ever happen.
This actually also just some safety logic to ensure that a vmo exists + is part of the worker status when the machineclasses are generated.